LanguageFeatures/Records/type_inference_A01_t02.dart

```diff
@@ -0,0 +1,43 @@
+// Copyright (c) 2022, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion Given a type schema K and a record expression E of the general
+/// form (e1, ..., en, d1 : e{n+1}, ..., dm : e{n+m}) inference proceeds as
+/// follows.
+///
+/// If K is a record type schema of the form
+/// (K1, ..., Kn, {d1 : K{n+1}, ...., dm : K{n+m}}) then:
+///
+/// - Each ei is inferred with context type schema Ki to have type Si
+///   - Let Ri be the greatest closure of Ki
+///   - If Si is a subtype of Ri then let Ti be Si
+///   - Otherwise, if Si is dynamic, then we insert an implicit cast on ei to
+///     Ri, and let Ti be Ri
+///   - Otherwise, if Si is coercible to Ri (via some sequence of call method
+///     tearoff or implicit generic instantiation coercions), then we insert the
+///     appropriate implicit coercion(s) on ei. Let Ti be the type of the
+///     resulting coerced value (which must be a subtype of Ri, possibly proper).
+///   - Otherwise, it is a static error.
+///  - The type of E is (T1, ..., Tn, {d1 : T{n+1}, ...., dm : T{n+m}})
+/// If K is any other type schema:
+///
+/// - Each ei is inferred with context type schema _ to have type Ti
+/// - The type of E is (T1, ..., Tn, {d1 : T{n+1}, ...., dm : T{n+m}})
+///
+/// @description Checks type inference for records.
+/// @author sgrekhov22@gmail.com
+
+// SharedOptions=--enable-experiment=records
+
+main() {
+  (num,) r1 = (42,);
+  r1.$0.isOdd;
```

Comment:
> The type of 42 is int (we do run type inference on that expression, but it doesn't do anything).
> 
> Then the getter $0 of r1 has static return type num, because the static type of r1 in (num,), and this means that r1.$0.isOdd should be a compile-time error.

But this contradicts  the [Type inference](https://github.com/dart-lang/language/blob/master/accepted/future-releases/records/records-feature-specification.md#type-inference) part of the spec

> As noted above, contrary to the practice for runtime checked covariant nominal types, we do not prefer the context type over the more precise upwards type. The following example illustrates this:
> ```dart
>   // No static error.
>   // Inferred type of the record is (int, double)
>   (num, num) r = (3, 3.5)..$0.isEven;
> ```

And this test tests this statement from the spec. Did I miss anything?

Reply:
Thank you for detailed explanation! I updated tests accordingly.